### ISSUE TITLE
Response bodies disappear proxying PUT requests

### ISSUE THREAD

**Issue by meare** [bbyars/mountebank#132]
The problem is that mountebank (1.4.3) seem to strip bodies from responses proxying PUT requests.
Imposter is pretty generic:

```
{
  "protocol": "http",
  "port": 4545,
  "name": "someservice",
  "requests": [],
  "stubs": [
    {
      "responses": [
        {
          "proxy": {
            "to": "http://someservice",
            "mode": "proxyAlways",
            "predicateGenerators": [
              {
                "matches": {
                  "method": true,
                  "path": true,
                  "query": true
                }
              }
            ]
          }
        }
      ]
    }
  ]
}
```

Service returns JSON body when it's called directly:

```
root@55ab5524a2fa:/# curl -X PUT -i someservice/db        
HTTP/1.1 401 Unauthorized
Server: nginx/1.8.0
Date: Fri, 11 Mar 2016 14:35:44 GMT
Content-Type: text/html; charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive

{"status":401,"msg":"Unauthorized"}
```

However mountebank seem to strip response body after proxying:

```
root@55ab5524a2fa:/# curl -X PUT -i localhost:4545/db
HTTP/1.1 401 Unauthorized
Server: nginx/1.8.0
Date: Fri, 11 Mar 2016 14:34:56 GMT
Content-Type: text/html; charset=UTF-8
Transfer-Encoding: chunked
Connection: close
```

I'm not pasting updated imposter but response 'body' field is empty (equal to "") there too.
It works as as expected with GET requests.
It's is really weird, because this proxy worked properly few months ago in older mountebank version (1.4.1 I think).
Thank you!

**Comment by bbyars** [comment]
Hi there,
Thanks for reporting.  I tried to reproduce a couple different ways and have been unable to.  Can you help me figure out how to reproduce?

The most like-for-like way I tried was to create the following http server (in file test.js):

    var http = require('http');

    var server = http.createServer(function (request, response) {
        response.writeHead(401, {
            Server: 'nginx/1.8.0',
            'Content-Type': 'text/html; charset=UTF-8'
        });
        response.end('{"status":401,"msg":"Unauthorized"}');
    });

    server.listen(3000, function () {
        console.log('Server listening on: http://localhost:3000');
    });

I ran this with `node test.js` to start the server, then confirmed that `curl -X PUT -i http://localhost:3000/db` gave the expected response:

    HTTP/1.1 401 Unauthorized
    Server: nginx/1.8.0
    Content-Type: text/html; charset=UTF-8
    Date: Fri, 11 Mar 2016 23:28:44 GMT
    Connection: keep-alive
    Transfer-Encoding: chunked

    {"status":401,"msg":"Unauthorized"}

I set up the same imposter configuration you provided, except to change the origin domain to localhost:3000:

    {
      "protocol": "http",
      "port": 4545,
      "name": "someservice",
      "requests": [],
      "stubs": [
        {
          "responses": [
            {
              "proxy": {
                "to": "http://someservice",
                "mode": "proxyAlways",
                "predicateGenerators": [
                  {
                    "matches": {
                      "method": true,
                      "path": true,
                      "query": true
                    }
                  }
                ]
              }
            }
          ]
        }
      ]
    }

I started mountebank with `mb --configfile test-imposter.js` and ran the curl command you show:

    nabbyars:~:% curl -X PUT -i localhost:4545/db
    HTTP/1.1 401 Unauthorized
    Server: nginx/1.8.0
    Content-Type: text/html; charset=UTF-8
    Date: Fri, 11 Mar 2016 23:24:14 GMT
    Connection: close
    Transfer-Encoding: chunked

    {"status":401,"msg":"Unauthorized"}%

The body shows up.  I don't spot the difference between what I did and what you did - any idea?

**Comment by bbyars** [comment]
For my own reference in case this turns out to be real, here's the test I wrote:

    promiseIt.only('should maintain response bodies proxying PUT requests (issue #132)', function () {
        var originServerPort = port + 1,
            originServerStub = { responses: [{ is: {
                statusCode: 401,
                body: { status: 401, msg: 'Unauthorized' }
            } }] },
            originServerRequest = {
                protocol: 'http',
                port: originServerPort,
                stubs: [originServerStub],
                name: this.name + ' origin'
            },
            proxyStub = { responses: [{ proxy: {
                to: 'http://localhost:' + originServerPort,
                mode: 'proxyAlways',
                predicateGenerators: [{
                    matches: {
                        method: true,
                        path: true,
                        query: true
                    }
                }]
            } }] },
            proxyRequest = { protocol: 'http', port: port, stubs: [proxyStub], name: this.name + ' proxy' };

        return api.post('/imposters', originServerRequest).then(function (response) {
            assert.strictEqual(response.statusCode, 201, JSON.stringify(response.body, null, 2));
            return api.post('/imposters', proxyRequest);
        }).then(function (response) {
            assert.strictEqual(response.statusCode, 201, JSON.stringify(response.body, null, 2));
            return client.put('/db', 'TEST', port);
        }).then(function (response) {
            assert.strictEqual(response.statusCode, 401);
            assert.deepEqual(JSON.parse(response.body), { status: 401, msg: 'Unauthorized' });
        }).finally(function () {
            return api.del('/imposters');
        });
    });

**Comment by sunfreakz** [comment]
Hi Brandon, thanks for a quick reply!
I couldn't reproduce this at home too, but here is what i finally figured out at work.
nginx was returning 0-byte-long responses for requests done via mb proxy. I used tcpdump to see what's the difference between direct and proxied requests.

Direct request:
```
PUT /db2 HTTP/1.1
User-Agent: curl/7.38.0
Host: someservice
Accept: */*
```
mountebank request:
```
PUT /db2 HTTP/1.1
User-Agent: curl/7.38.0
host: someservice
Accept: */*
connection: close
Transfer-Encoding: chunked
```
It appears that mb is doing chunked requests, so did I via curl:
```
root@6e01ee040e9d:/# curl -X PUT -i -H "Transfer-encoding: chunked" -d "" someservice/db2
HTTP/1.1 201 Created
Server: nginx/1.8.0
Date: Tue, 15 Mar 2016 11:29:55 GMT
Content-Type: text/html;charset=UTF-8
Transfer-Encoding: chunked
Connection: keep-alive

curl: (18) transfer closed with outstanding read data remaining
```
Adding Content-length header to response did not solved this curl error.
So cause of the problem are chunked requests that proxy sends and, apparently, badly configured server that does not respond properly :)

**Comment by bbyars** [comment]
Ah that helps tremendously, thanks!

**Comment by bbyars** [comment]
I think this is fixed as of commit af8302c. Will release soon.

**Comment by DeivDave** [comment]
Hi, I have a question, how can I do for effecting a mock through proxy mountebank? That is I would like it, that if I wrote localhost:4545  was shown me the page of google.com, is this possible?
Thanks thousand to everybody
